pkg/compactor/blocks_cleaner.go

```diff
@@ -391,3 +408,50 @@ func (c *BlocksCleaner) cleanUserPartialBlocks(ctx context.Context, partials map
 		level.Info(userLogger).Log("msg", "deleted partial block marked for deletion", "block", blockID)
 	}
 }
+
+// applyUserRetentionPeriod marks blocks for deletion which have aged past the retention period.
+func (c *BlocksCleaner) applyUserRetentionPeriod(ctx context.Context, idx *bucketindex.Index, userID string, userBucket *bucket.UserBucketClient, userLogger log.Logger) {
+	retention := c.cfgProvider.CompactorRetentionPeriod(userID)
+	now := time.Now()
+
+	level.Debug(userLogger).Log("msg", "applying retention", "retention", retention.String())
+	blocks := listBlocksOutsideRetentionPeriod(idx, now, retention)
+
+	// Attempt to mark all blocks. It is not critical if a marking fails, as
+	// the cleaner will retry applying the retention in its next cycle.
+	for _, b := range blocks {
+		level.Info(userLogger).Log("msg", "applied retention: marking block for deletion", "block", b.ID, "maxTime", b.MaxTime)
+		if err := block.MarkForDeletion(ctx, userLogger, userBucket, b.ID, fmt.Sprintf("block exceeding retention of %v", retention), c.blocksMarkedForDeletion); err != nil {
+			level.Warn(userLogger).Log("msg", "failed to mark block", "block", b.ID, "err", err)
```

Comment:
We should mention "mark for what..." (eg. in Thanos there's also a marker for "no compaction").

```suggestion
			level.Warn(userLogger).Log("msg", "failed to mark block for deletion", "block", b.ID, "err", err)
```